crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+            let chain = self.chains.remove(&chain_id).expect("To fork to be present");
+            block_fork = chain.fork_block();
+            let (canonical, rest) = chain.split_at_number(block_fork_number);
+            let canonical = canonical.expect("Chain is present");
+            // reinsert back the chunk of sidechain that didn't get reorged.
+            if let Some(rest_of_sidechain) = rest {
+                self.block_indices.insert_chain(chain_id, &rest_of_sidechain);
+                self.chains.insert(chain_id, rest_of_sidechain);
+            }
+            block_fork_number = canonical.fork_block_number();
+            chains_to_promote.push(canonical);
+        }
+
+        let old_tip = self.block_indices.canonical_tip();
+        // Merge all chain into one chain.
+        let mut new_canon_chain = chains_to_promote.pop().expect("There is at least one block");
+        for chain in chains_to_promote.into_iter().rev() {
+            new_canon_chain.append_chain(chain);
+        }
+
+        // update canonical index
+        self.block_indices.canonicalize_blocks(&new_canon_chain.blocks);
+
+        // if joins to the tip
+        if new_canon_chain.fork_block_hash() == old_tip.hash {
+            // append to database
+            self.commit_canonical(new_canon_chain)?;
+        } else {
+            // it forks to canonical block that is not the tip.
+
+            let canon_fork = new_canon_chain.fork_block();
+            // sanity check
+            if self.block_indices.canonical_hash(&canon_fork.number) != Some(canon_fork.hash) {
+                unreachable!("all chains should point to canonical chain.");
+            }
+
+            // revert `N` blocks from current canonical chain and put them inside BlockchanTree
+            // This is main reorgs on tables.
+            let old_canon_chain = self.revert_canonical(canon_fork.number)?;
+            self.commit_canonical(new_canon_chain)?;
+
+            // insert old canonical chain to BlockchainTree.
+            // TODO check if there there is chains that can be merged.
+            self.insert_chain(old_canon_chain);
+        }
+
+        Ok(())
+    }
+
+    /// Commit chain for it to become canonical. Assume we are doing pending operation to db.
+    fn commit_canonical(&mut self, chain: Chain) -> Result<(), Error> {
+        let mut tx = Transaction::new(&self.externals.db)?;
+
+        let new_tip = chain.tip().number;
+
+        for item in chain.blocks.into_iter().zip(chain.changesets.into_iter()) {
+            let ((_, block), changeset) = item;
+            tx.insert_block(block, self.externals.chain_spec.as_ref(), changeset).map_err(|e| {
+                println!("commit error:{e:?}");
+                ExecError::VerificationFailed
+            })?;
+        }
+        // update pipeline progress.
+        tx.update_pipeline_stages(new_tip).map_err(|_| ExecError::VerificationFailed)?;
+
+        // TODO error cast
```

Comment:
Is this still a todo?

Reply:
Refactoring errors will probably help: https://github.com/paradigmxyz/reth/issues/1458

Will add `String` error to have the needed information.